src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java

```diff
@@ -0,0 +1,158 @@
+
+import com.cloudbees.plugins.credentials.CredentialsMatchers;
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.cloudbees.plugins.credentials.common.StandardListBoxModel;
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import com.cloudbees.plugins.credentials.common.UsernamePasswordCredentials;
+import com.cloudbees.plugins.credentials.domains.URIRequirementBuilder;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.*;
+import hudson.scm.SCM;
+import hudson.scm.SCMDescriptor;
+import hudson.security.ACL;
+import hudson.slaves.WorkspaceList;
+import hudson.util.ListBoxModel;
+import jenkins.model.Jenkins;
+import org.apache.commons.httpclient.methods.GetMethod;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.workflow.cps.persistence.PersistIn;
+import org.jenkinsci.plugins.workflow.flow.*;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.Stapler;
+import org.kohsuke.stapler.StaplerRequest;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import static org.jenkinsci.plugins.workflow.cps.persistence.PersistenceContext.JOB;
+
+@PersistIn(JOB) public class CpsHttpFlowDefinition extends FlowDefinition {
+
+    private final String scriptUrl;
+    private final String credentialsId;
+    private final int retryCount;
+
+    @DataBoundConstructor public CpsHttpFlowDefinition(String scriptUrl, String credentialsId, int retryCount) {
+        this.scriptUrl = scriptUrl.trim();
+        this.credentialsId = credentialsId;
+        this.retryCount = retryCount;
+    }
+
+    public String getCredentialsId() {
+        return credentialsId;
+    }
+
+    public String getScriptUrl() {
+        return scriptUrl;
+    }
+
+    public int getRetryCount() {
+        return retryCount;
+    }
+
+    @Override public CpsFlowExecution create(FlowExecutionOwner owner, TaskListener listener, List<? extends Action> actions)
+            throws Exception {
+        Queue.Executable _build = owner.getExecutable();
+        if (!(_build instanceof Run)) {
+            throw new IOException("Can only pull a Jenkinsfile in a run");
+        }
+        Run<?, ?> build = (Run<?, ?>) _build;
+
+        String expandedScriptUrl = build.getEnvironment(listener).expand(scriptUrl);
+        URL url = new URL(expandedScriptUrl);
+        listener.getLogger().println("Fetching pipeline from " + expandedScriptUrl);
+        int count = 0;
+        int maxTries = retryCount + 1;
+
+        while (true) {
+            try {
+                HttpURLConnection con = (HttpURLConnection) url.openConnection();
```

Comment:
https://github.com/jenkinsci/apache-httpcomponents-client-4-api-plugin/blob/master/src/main/java/io/jenkins/plugins/httpclient/RobustHTTPClient.java could be used for better stability

Reply:
And easier authentication probably.

Reply:
I haven't found any usage of this anywhere. Do you have any examples?

Reply:
[query](https://github.com/search?q=user%3Ajenkinsci+RobustHTTPClient&type=Code)